test/Microsoft.AspNetCore.Routing.Tests/Matching/DfaMatcherTest.cs

```diff
@@ -157,6 +164,173 @@ public async Task MatchAsync_MultipleMatches_EndpointSelectorCalled()
             Assert.Equal(endpoint2, endpointFeature.Endpoint);
         }
 
+        [Fact]
+        public async Task MatchAsync_NoCandidates_Logging()
+        {
+            // Arrange
+            var endpointDataSource = new DefaultEndpointDataSource(new List<Endpoint>
+            {
+                CreateEndpoint("/{p:int}", 0)
+            });
+
+            var sink = new TestSink();
+            var matcher = CreateDfaMatcher(endpointDataSource, loggerFactory: new TestLoggerFactory(sink, enabled: true));
+
+            var (httpContext, endpointFeature) = CreateHttpContext();
+            httpContext.Request.Path = "/";
+
+            // Act
+            await matcher.MatchAsync(httpContext, endpointFeature);
+
+            // Assert
+            Assert.Null(endpointFeature.Endpoint);
+
+            Assert.Collection(
+                sink.Writes,
+                (log) =>
+                {
+                    Assert.Equal(DfaMatcher.EventIds.CandidatesNotFound, log.EventId);
+                    Assert.Equal("No candidates found for the request path '/'", log.Message);
+                });
+        }
+
+        [Fact]
+        public async Task MatchAsync_ConstraintRejectsEndpoint_Logging()
+        {
+            // Arrange
+            var endpointDataSource = new DefaultEndpointDataSource(new List<Endpoint>
+            {
+                CreateEndpoint("/{p:int}", 0)
+            });
+
+            var sink = new TestSink();
+            var matcher = CreateDfaMatcher(endpointDataSource, loggerFactory: new TestLoggerFactory(sink, enabled: true));
+
+            var (httpContext, endpointFeature) = CreateHttpContext();
+            httpContext.Request.Path = "/One";
+
+            // Act
+            await matcher.MatchAsync(httpContext, endpointFeature);
+
+            // Assert
+            Assert.Null(endpointFeature.Endpoint);
+
+            Assert.Collection(
+                sink.Writes,
+                (log) => 
+                {
+                    Assert.Equal(DfaMatcher.EventIds.CandidatesFound, log.EventId);
+                    Assert.Equal("1 candidate(s) found for the request path '/One'", log.Message);
+                },
+                (log) =>
+                {
+                    Assert.Equal(DfaMatcher.EventIds.CandidateRejectedByConstraint, log.EventId);
+                    Assert.Equal("Endpoint '/{p:int}' with route pattern '/{p:int}' was rejected by constraint 'p' with value 'One' for the request path '/One'", log.Message);
```

Comment:
p is the parameter name, not constraint name.

Did you mean to say "...constraint on parameter 'p'..."? FYI non-inline constraints won't have an inline name. Maybe include the type name of the constraint?

Reply:
I said constraint name for terseness - do you think there's a way to make it clearer without so many words?. non-inline constraints definitely have a name.

Reply:
I guess I dunno what you mean then. I don't know what an inline prefix is.

Reply:
These things - https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/RouteOptions.cs#L52-L76

Reply:
OK, I think I get what you mean now. I was thinking that logging the parameter name is enough. since you generally you wrote the route so you should know what's going on. Maybe we should log the `ToString()` of the constraint instance? 

By default that will be the type name, but if we had a case we cared about we could customize it.

Reply:
I still think this log message is questionable. p is the parameter name. Saying the value was rejected by constraint 'p' doesn't make sense to me.

Also what happens if a parameter has multiple constraints on it. The logging doesn't give you an indication of what constraint it was that failed.

Reply:
So I ended up pushing this with broken tests becuase I'm a silly billy. Now it looks like this:

> "Endpoint '/{p:int}' with route pattern '/{p:int}' was rejected by constraint 'p':'Microsoft.AspNetCore.Routing.Constraints.IntRouteConstraint' with value 'One' for the request path '/One'"

Does that address your concerns? If it doesn't please make a suggestion.

Reply:
:shipit:

Reply:
Cool beans!